include/jemalloc/internal/thread_event.h

```diff
@@ -0,0 +1,114 @@
+#ifndef JEMALLOC_INTERNAL_THREAD_EVENT_H
+#define JEMALLOC_INTERNAL_THREAD_EVENT_H
+
+#include "jemalloc/internal/tsd.h"
+
+/*
+ * Maximum threshold on thread_allocated_threshold_fast, so that there is no
+ * need to check overflow in malloc fast path. (The allocation size in malloc
+ * fast path never exceeds SC_LOOKUP_MAXCLASS.)
+ */
+#define THREAD_ALLOCATED_THRESHOLD_FAST_MAX				\
+    (UINT64_MAX - SC_LOOKUP_MAXCLASS + 1)
+
+/*
+ * List of all waits, in the following format:
+ *  W(event,		(condition))
+ */
+#define ITERATE_OVER_ALL_WAITS						\
+    W(prof_sample,	(config_prof && opt_prof))
+
+void thread_event_trigger(tsd_t *tsd, uint64_t usize, bool delay_event);
+void thread_event_update(tsd_t *tsd);
+
+JEMALLOC_ALWAYS_INLINE uint64_t
+thread_event_min_wait_helper(tsd_t *tsd) {
+	uint64_t wait = UINT64_MAX;
+
+#define W(event, condition)						\
+	if (condition) {						\
+		uint64_t event_wait =					\
+		    tsd_##event##_event_wait_get(tsd);			\
+		if (event_wait > 0 && event_wait < wait) {		\
+			wait = event_wait;				\
+		}							\
+	}
+
+	ITERATE_OVER_ALL_WAITS
+#undef W
+
+	return wait;
+}
+
+JEMALLOC_ALWAYS_INLINE void
+thread_event_assert_invariants(tsd_t *tsd) {
+	if (!config_debug) {
+		return;
+	}
+
+	uint64_t thread_allocated = tsd_thread_allocated_get(tsd);
+	uint64_t last_event = tsd_thread_allocated_last_event_get(tsd);
+	uint64_t threshold = tsd_thread_allocated_threshold_get(tsd);
+	uint64_t threshold_fast = tsd_thread_allocated_threshold_fast_get(tsd);
+
+	assert(last_event != threshold);
+	if (likely(threshold <= THREAD_ALLOCATED_THRESHOLD_FAST_MAX)) {
+		assert(threshold_fast == threshold);
+	} else {
+		assert(threshold_fast == 0);
+	}
+
+	/* Both subtractions are intentionally susceptible to underflow. */
+	assert(thread_allocated - last_event < threshold - last_event);
+
+	/*
+	 * The assert for the min wait time can only fail (trivially) when both
+	 * of the following two conditions hold:
+	 * (1) the TSD is just initialized, and
```

Comment:
Let's call thread_event() (or a special tsd_thread_event_init if necessary) at the end of TSD initialization? It will have no event to record, but can help put everything in place? It looks right now when all events are off (config_prof && !opt_prof), it goes through medium path instead of fast.

Reply:
> Let's call thread_event() (or a special tsd_thread_event_init if necessary) at the end of TSD initialization? It will have no event to record, but can help put everything in place?

By "put everything in place" what do you mean? the `thread_event()` call will zero the wait time but won't fetch the next wait times, which relies on the logic of each individual event. If we want something more on `tsd` initialization, e.g. some kind of `tsd_thread_event_init()`, to also properly fetch the next wait times, then it might do a lot more: to fetch the prof sampling wait time, we need to initialize the `tdata`, but our current logic has a lazy `tdata` init approach - we won't initialize it until the time of the first real sample.

> It looks right now when all events are off (config_prof && !opt_prof), it goes through medium path instead of fast.

Yes it goes through the medium path, but only once. After that, the wait times are properly set, and in case where all events are off, the threshold is set to max.

Reply:
> By "put everything in place" what do you mean? 

If it helps to have an init function, we can add one. e.g. if it helps simplify the thread_event logic and makes sure some relation always holds between threshold and allocated etc (instead of assuming we have to tolerate the all zero case).

> Yes it goes through the medium path, but only once. 

I tried the stress benchmark and it didn't look so. malloc_default was the common path for me.

Reply:
For now the stress test always falls to medium path: it has `opt_prof` off, so threshold is set to `UINT64_MAX` which rolls back, and `thread_allocated` is always larger than `thread_allocated_threshold_fast`. We can either try to manually avoid rolling back the threshold for now, or wait until the GC counter is added, in which case the `UINT64_MAX` scenario will never happen. Thoughts?

Reply:
Discussed offline. We'll optimize for the opt_prof off case.